doc/bulk_counter/bulk_counter.md

```diff
@@ -0,0 +1,144 @@
+# SONiC Bulk Counter Design #
+
+## Table of Content
+
+### Revision
+
+ | Rev |     Date    |       Author       | Change Description                |
+ |:---:|:-----------:|:------------------:|-----------------------------------|
+ | 0.1 |             |      Junchao Chen  | Initial version                   |
+
+### Scope
+
+This document is the design document for bulk counter feature on SONiC.
+
+### Definitions/Abbreviations
+
+N/A
+
+### Overview
+
+PR https://github.com/opencomputeproject/SAI/pull/1352/files introduced new SAI APIs that supports bulk stats:
+
+- sai_bulk_object_get_stats
+- sai_bulk_object_clear_stats
+
+SONiC flex counter infrastructure shall utilize bulk stats API to gain better performance. This document discusses how to integrate these two new APIs to SONiC. 
+
+### Requirements
+
+- Syncd shall use bulk stats APIs based on object type. E.g. for a counter group that queries queue and pg stats, queue stats support bulk while pg stats does not, in that case queue stats shall use bulk API, pg stats shall use non bulk API
+- For a certain object type in a counter group, it shall use bulk stats only if:
+  - The stats capability for each counter IDs shall include SAI_STATS_MODE_BULK_READ, SAI_STATS_MODE_BULK_CLEAR, SAI_STATS_MODE_BULK_READ_AND_CLEAR (avoid handling stats mode change)
+  - Each object queries exactly the same counter IDs. (Requirement from function signature of sai_bulk_object_get_stats and sai_bulk_object_clear_stats)
+- Syncd shall automatically fall back to old way if bulk stats APIs are not supported
+- Syncd shall utilize API sai_query_stats_capability to query bulk capability. Syncd shall treat counter as no bulk capability if API sai_query_stats_capability return error except SAI_STATUS_BUFFER_OVERFLOW (SAI_STATUS_BUFFER_OVERFLOW requires a retry with larger buffer)
+- Syncd shall call bulk stats API in flex counter thread and avoid calling it in main thread to make sure main thread only handles short and high priority tasks. (This is the default behavior in flex counter infrastructure)
+- In phase 1, the change is limited to syncd only, no CLI/swss change. Syncd shall use bulk stats API by querying the statistic capability. Syncd shall deduce the bulk stats mode according to the stats mode defined in FLEX DB:
+  - SAI_STATS_MODE_READ -> SAI_STATS_MODE_BULK_READ
+  - SAI_STATS_MODE_READ_AND_CLEAR -> SAI_STATS_MODE_BULK_READ_AND_CLEAR
+
+### Architecture Design
```

Comment:
Here `bulk get` shall be a block API and counter thread calls it to talk with asic/npu in sync mode. Is it possible to support async IO, i.e `bulk get` is non-block and provides callback, asic/npu will call callback to finish rest things after counter collection? Since bulk get collets a lot of data, async IO is much efficient. With async IO, another benefit is we do not need many counter threads.

Reply:
Yes, `bulk get` is a block API. Async IO is a great idea, and it will surely improve the performance. It requires  several new SAI API first and might be another feature.